### PR TITLE
Adds support custom name resolver in codegen

### PR DESCRIPTION
**Pull request by negezor**
I added the `NameResolver` trait for better maintainability. I haven't touched columns and other parts that don't depend on module names yet, and I haven't deleted the getter code either.

## Adds

- Adds a new argument to sea-orm-cli `--singularize`, which generates singular module names.
- Create your own name transformations in `sea-orm-codegen`

## Breaking Changes

- Requires one more argument in `EntityTransformer::transform`

### CODE REVIEW

File: sea-orm-codegen/src/entity/name_resolver.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    fn resolve_entity_name(&self, name: &str) -> String {
+        singular(name.to_camel_case())
+    }
+
+    fn resolve_relation_name(&self, name: &str) -> String {
+        singular(name.to_camel_case())
+    }
+}
```
Comment (billy1624) [inline sea-orm-codegen/src/entity/name_resolver.rs:57]: Having a name resolver is a good idea :)
Can we supply the output of `NameResolver`'s default implementation to `singular` method?

Comment (negezor) [reply sea-orm-codegen/src/entity/name_resolver.rs:57]: @billy1624 yes of course

**Review by tyt2y3 — COMMENTED** [review]
Actually, we don't need a trait here. Everything can be done by a struct `NameResolver` and it only needs to have a state `singularize: bool` which can only be configured in the constructor `NameResolver::new()`

This is better that the last PR indeed

File: sea-orm-cli/src/commands.rs [hunk 2]
```diff
@@ -150,7 +151,9 @@ pub async fn run_generate_command(matches: &ArgMatches<'_>) -> Result<(), Box<dy
                 _ => unimplemented!("{} is not supported", url.scheme()),
             };
 
-            let output = EntityTransformer::transform(table_stmts)?
+            let name_resolver = NameResolver::new(singularize);
+
+            let output = EntityTransformer::transform(table_stmts, name_resolver)?
                 .generate(expanded_format, WithSerde::from_str(with_serde).unwrap());
```
Comment (billy1624) [inline sea-orm-cli/src/commands.rs:157]: Hey @negezor, thanks for the updates! I think the semantic isn't seems right. The transformer, `EntityTransformer::transform(table_stmts, name_resolver)`, don't need name resolver. Instead, we could have introduce a new struct call `EntityWriterContext` which contains three things...

```rs
pub struct EntityWriterContext {
  pub(crate) expanded_format: bool,
  pub(crate) with_serde: WithSerde,
  pub(crate) name_resolver: NameResolver,
}
```

Then, `EntityWriter::generate` method would take an `EntityWriterContext`. Thoughts?

File: sea-orm-codegen/src/entity/base_entity.rs [hunk 3]
```diff
@@ -125,6 +159,27 @@ impl Entity {
...
+            .collect()
+    }
+
+    pub fn resolve_conjunct_relations_to_relation_name(&self, name_resolver: &NameResolver) -> Vec<Ident> {
+        self.conjunct_relations
+            .iter()
+            .map(|con_rel| con_rel.resolve_to_relation_name(name_resolver))
+            .collect()
+    }
+
```
Comment (billy1624) [inline sea-orm-codegen/src/entity/base_entity.rs:182]: All these `resolve_*` methods (here, and in other files except those in `name_resolver.rs`) seems just a duplicate of the original methods few line below, e.g. `resolve_conjunct_relations_via_module_name` method is just `get_conjunct_relations_via_snake_case` method with name resolve applied. We should change the logic in the original method, i.e. applying name resolver in it.

File: sea-orm-cli/src/commands.rs (hunk 2 above)
Comment (tyt2y3) [reply sea-orm-cli/src/commands.rs:157]: I agree indeed

Comment (nahuakang) [reply sea-orm-cli/src/commands.rs:157]: @billy1624 In the case where we discussed about saving a context info on which datetime crate the user chooses in [the `--date-time-crate` flag for `sea-orm-cli`](https://github.com/SeaQL/sea-orm/issues/661), we'd basically write:
```rust
pub struct EntityWriterContext {
  pub(crate) expanded_format: bool,
  pub(crate) with_serde: WithSerde,
  pub(crate) name_resolver: NameResolver,
  pub(crate) date_time_crate: &str,
}
```
and pass this context all the way down to `Column.get_rs_type()` for resolving the Rust types based on `chrono` vs `time`? :)

**Comment by billy1624** [comment]
Hey @negezor, any updates? And do you need any help? :)

**Comment by billy1624** [comment]
> Also, since in both of our PRs we're using `EntityWriterContext`, maybe we could merge this PR and [PR#724](https://github.com/SeaQL/sea-orm/pull/724) sequentially? I'm happy to help with the context update, too.

Good suggestions! We could consolidate all flags into `EntityWriterContext`.

**Comment by negezor** [comment]
Sorry for the long answer. Now this PR depends on #724.

File: sea-orm-codegen/src/entity/base_entity.rs (hunk 3 above)
Comment (negezor) [reply sea-orm-codegen/src/entity/base_entity.rs:182]: I didn't touch them originally, as that would have required more changes. Also, they are now more than just getters. Perhaps this one could be changed by later refactoring. 🤔

**Review by billy1624 — COMMENTED** [review]
Hey @negezor, sorry for the delay. Since https://github.com/SeaQL/sea-orm/pull/724 has been merged. Would you like to rebase?

**Comment by negezor** [comment]
I think I did the rebase wrong somehow 😅

File: sea-orm-codegen/src/entity/name_resolver.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    pub fn resolve_relation_name(&self, name: &str) -> String {
+        let name = name.to_camel_case();
+
+        if self.singularize {
+            singular(name)
+        } else {
+            name
+        }
+    }
```
Comment (billy1624) [inline sea-orm-codegen/src/entity/name_resolver.rs:44]: Hey @negezor, thanks for the hard rebase (it's hard as this PR is far behind from master).

I wonder why we need three different methods? I think we can merge it and have two methods? One for snake_case and the other for camel_case.

Comment (negezor) [reply sea-orm-codegen/src/entity/name_resolver.rs:44]: A fair remark, but I proceeded from the clear goals of the method. I can make changes if necessary.

**Pull request closed without merging**